packages/financial-templates-lib/src/proxy-transaction-handler/DSProxyManager.js

```diff
@@ -0,0 +1,150 @@
+const { getFromBlock } = require("@uma/common");
+
+class DSProxyManager {
+  /**
+   * @notice Constructs new Liquidator bot.
+   * @param {Object} logger Module used to send logs.
+   * @param {Object} web3 Web3 object to submit transactions and process on-chain info.
+   * @param {Object} gasEstimator Module used to estimate optimal gas price with which to send txns.
+   * @param {String} account Ethereum account from which to send txns.
+   * @param {String} dsProxyFactoryAddress address of the DSProxy factory to create new DSProxies and the like.
+   */
+  constructor({ logger, web3, gasEstimator, account, dsProxyFactoryAddress, dsProxyFactoryAbi, dsProxyAbi }) {
+    assert(web3.utils.isAddress(account), "Account needs to be a valid address");
+    assert(web3.utils.isAddress(dsProxyFactoryAddress), "dsProxyFactoryAddress needs to be a valid contract address");
+    this.logger = logger;
+    this.account = account;
+    this.web3 = web3;
+    this.gasEstimator = gasEstimator;
+    this.dsProxyFactoryAddress = dsProxyFactoryAddress;
+    this.dsProxyFactory = new web3.eth.Contract(dsProxyFactoryAbi, dsProxyFactoryAddress);
+    this.dsProxyAbi = dsProxyAbi;
+    this.dsProxy = null;
+    this.dsProxyAddress = null;
+
+    // Helper functions from web3.
+    this.BN = this.web3.utils.BN;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.fromWei = this.web3.utils.fromWei;
+    this.utf8ToHex = this.web3.utils.utf8ToHex;
+    this.isAddress = this.web3.utils.isAddress;
+
+    // Multiplier applied to Truffle's estimated gas limit for a transaction to send.
+    this.GAS_LIMIT_BUFFER = 1.25;
+  }
+
+  getDSProxyFactoryAddress() {
+    return this.dsProxyFactoryAddress;
+  }
+
+  getDSProxyAddress() {
+    return this.dsProxyAddress;
+  }
+
+  // Load in an existing DSProxy for the account EOA if one already exists OR create a new one for the user.
+  async initializeDSProxy() {
+    this.logger.debug({
+      at: "DSProxyManager",
+      message: "Initalizing...Looking for existing DSProxies or deploying a new one for the provided EOA",
+      dsProxyFactoryAddress: this.dsProxyFactoryAddress
+    });
+
+    if (this.dsProxy && this.dsProxyAddress) this.dsProxyAddress;
+    const fromBlock = await getFromBlock(this.web3);
+    const events = await this.dsProxyFactory.getPastEvents("Created", {
+      fromBlock,
+      filter: { owner: this.account }
+    });
+
+    // The user already has a DSProxy deployed. Load it in from the events.
+    if (events.length > 0) {
```

Comment:
i feel like this function should be split up into multiple functions which handle parts of the logic
async hasProxy() => checks cache / chain to see if user has already deployed this
async getProxy() => loads cached proxy or pulls it from chain to cache. throws if does not exist.
async createProxy() => only creates if proxy does not exist, throws if it does

i think this kind of makes it easier to reason about and easier to unit test and these would be composed to create your init function. wdyt?

Reply:
ye we could do that. however I feel that seeing there is only two branching states this is probs fine to leave like this for now. I can add this refactor if you feel strongly about it but I dont think it's really warranted at this point. 

Also, I tried to implement this change but it ended up taking more code and had to repeat the log query method against the DSProxy factory. For now I think it cleaner to just leave it like this. In places where initialisation must happen first I added an assert to check that an address is set. I think this amounts to the same behaviour but with fewer async calls on chain and less code.

Reply:
ok yea sounds good. not super opinionated on it, just wanted to get your take